File: Autonomous Mobile Robots/ex11A_harmonic_potential_fields_solution/matlab/Q2_potential_field_solution.m
%% Autonomous Mobile Robots - Exercise 11 Q2 - Harmonic Potential Fields
close all;
clear all;
clc;

%% 2.1 - Initialisation

% initialize traversability map. 1 for traversable, 0 for obstacle
Map = ones(11,9);
Map(1,:) = 0; Map(11,:) = 0; Map(:,1) = 0; Map(:,9)     = 0;
Map(9,2) = 0; Map(10,2) = 0; Map(10,3)= 0; Map(5:6,5:8) = 0;

% initialize search start and goal locations
SearchStart = [3,7];
SearchGoal  = [9,6];

fh = figure(); fh.Position(3:4) = [800, 400];
ax1 = subplot(1, 2, 1); hold on;
title(ax1, 'Obstacle map');
imshow(Map);
plot(ax1, SearchStart(2), SearchStart(1), 'g^');
plot(ax1, SearchGoal(2), SearchGoal(1), 'ro');

% Initialize iterative search
SearchSolution = zeros(size(Map));
SearchSolution(Map==0)=1;   %set obstacle cells to "1" (high potential)
SearchSolution(Map==1) =0.5; %set free cells to "0.5" (mid potential)
SearchSolution(SearchGoal(1),SearchGoal(2)) = 0;

%% 2.2 Iterative updates
% Iterative parameters (max iterations, convergence tolerance)
tol      = 0.01;
maxIter  = 50;

ax2 = subplot(1, 2, 2); hold on;
title(ax2, 'Potential Field');
hImage = imshow(SearchSolution, 'Colormap', parula);

% Iteratively solve the discrete Laplace Equation with Dirichlet boundary conditions
iter = 0; maxChange = inf;
while maxChange > tol
    iter = iter+1;
    assert(maxIter > iter, 'maxIter assert triggered. Aborting.');

	NextSearchSolution = SearchSolution;
    for x=1:1:size(Map,1)
    	for y=1:1:size(Map,2)
        	if and(Map(x,y)==1, SearchSolution(x,y)~=0)
        		NextSearchSolution(x,y) = 1/4*(SearchSolution(x-1,y) + ...
        	                                   SearchSolution(x+1,y) + ...
        								       SearchSolution(x,y-1) + ...
        		                               SearchSolution(x,y+1) );
        	end
        end
    end
    
    % Convolution solution
    % NextSearchSolution = conv2(SearchSolution, [0, 0.25, 0; 0.25, 0, 0.25; 0, 0.25, 0], 'same');
    % NextSearchSolution(Map==0)=1;
    % NextSearchSolution(SearchGoal(1),SearchGoal(2)) = 0;
        
    maxChange = max(max(abs(SearchSolution-NextSearchSolution)));
    SearchSolution = NextSearchSolution;
    
    % Animate
    pause(0.2);
    set(hImage, 'CData', SearchSolution);
end

hold on;
plot(ax2, SearchStart(2), SearchStart(1), 'g^');
plot(ax2, SearchGoal(2), SearchGoal(1), 'ro');
      
%% 2.3 Backtrack solution
% extract solution path from start to goal
% Construct the path as an n x 2 array, with the start as the first row and
% the goal as the last row. Assume an 8-connected grid

OptimalPath = SearchStart;
iter = 0;
while ~isequal(OptimalPath(end,:),SearchGoal)
    iter = iter+1;
    assert(maxIter > iter, 'maxIter assert triggered. Aborting.');
        
    % extract index corresponding to the minimal
    % value within a 3 by 3 window
    Window = SearchSolution(OptimalPath(end,1)-1:OptimalPath(end,1)+1, ...
                            OptimalPath(end,2)-1:OptimalPath(end,2)+1);
    [x,y] = find(Window==min(min(Window)));
    NextPoint = OptimalPath(end,:)+[x(1)-2,y(1)-2];
    OptimalPath = [OptimalPath; NextPoint];
end
plot(ax2, OptimalPath(:,2), OptimalPath(:,1), 'r-')

File: Autonomous Mobile Robots/ex11A_harmonic_potential_fields_solution/python/Q2_potential_field_solution.py
## Autonomous Mobile Robots - Exercise 11 Q2 - Harmonic Potential Fields
import numpy as np
import matplotlib.pyplot as plt
# from scipy.signal import convolve2d        # <- For convolution solution

# 2.1 - Initialisation
# Initialize traversability map
# 1 for traversable, 0 for obstacle
map = np.ones((11, 9))
map[:, [0, -1]] = 0
map[[0, -1], :] = 0
map[4:6, 4:8] = 0
map[8:10, 1] = 0
map[9, 2] = 0

# Initialize search start and goal locations
search_start = [2, 6]
search_goal = [8, 5]

fh, ah = plt.subplots(1, 2)
fh.set_size_inches(8, 5)

ah[0].set_title('Obstacle map')
ah[0].imshow(map, aspect='equal', cmap='gray')
ah[0].plot(search_start[1], search_start[0], 'g^')
ah[0].plot(search_goal[1], search_goal[0], 'ro')

# Initialize iterative search
potential_field = np.zeros_like(map)
potential_field[map == 0] = 1  # set obstacle cells to 1 (high potential)
potential_field[map == 1] = 0.5  # set free cells to 0.5 (mid potential)
potential_field[search_goal[0], search_goal[1]] = 0  # set goal potential

## 2.2 Iterative updates
# Iterative parameters(max iterations, convergence tolerance)
tol = 0.01
maxIter = 50

ah[1].set_title('Potential Field')
h_image = ah[1].imshow(potential_field, aspect='equal', cmap='viridis')

# Iteratively solve the discrete Laplace equation with
# Dirichlet boundary conditions
i = 0
max_change = 1.0

while max_change > tol:
    assert i < maxIter, 'maxIter assert triggered. Aborting.'
    i += 1

    next_potential = potential_field.copy()

    for x in range(map.shape[0]):
        for y in range(map.shape[1]):
            if map[x, y] == 1 and potential_field[x, y] != 0:
                next_potential[x, y] = 0.25 * (potential_field[x - 1, y] +
                                               potential_field[x + 1, y] +
                                               potential_field[x, y - 1] +
                                               potential_field[x, y + 1])

    # Alternative convolution:
    # next_potential = convolve2d(potential_field, np.array([[0, 0.25, 0], [0.25, 0, 0.25], [0, 0.25, 0]]), mode='same', fillvalue=1.0)
    # next_potential[map == 0] = 1
    # next_potential[search_goal[0], search_goal[1]] = 0

    max_change = np.abs(potential_field - next_potential).max()
    potential_field = next_potential

    # Animate
    plt.pause(0.2)
    h_image.set_data(potential_field)

ah[1].plot(search_start[1], search_start[0], 'g^')
ah[1].plot(search_goal[1], search_goal[0], 'ro')

## 2.3 Backtrack solution
# Extract solution path from start to goal
# Construct the path as a list of points, from the start to the goal
# Assume an 8-connected grid

optimal_path = [search_start]
i = 0
while optimal_path[-1] != search_goal:
    assert i < maxIter, 'maxIter assert triggered. Aborting.'
    i += 1

    # extract index corresponding to the minimal value within a 3x3 window
    cx, cy = optimal_path[-1]
    x, y = np.unravel_index(
        np.argmin(potential_field[cx - 1:cx + 2, cy - 1:cy + 2]), (3, 3))
    optimal_path.append([cx + x - 1, cy + y - 1])

ah[1].plot([y for x, y in optimal_path], [x for x, y in optimal_path], 'r-')
plt.show()
